Control/Concurrent/MVar/Strict.hs

```diff
@@ -88,68 +96,31 @@ newMVar value =
     putMVar mvar value  >>
     return mvar
 
-{-|
-  This is a combination of 'takeMVar' and 'putMVar'; ie. it takes the value
-  from the 'MVar', puts it back, and also returns it.
--}
-readMVar :: NFData a => MVar a -> IO a
-readMVar m = mask $ \_ -> do
-    a <- takeMVar m
-    putMVar m a
-    return a
-
 {-|
   Take a value from an 'MVar', put a new value into the 'MVar' and
   return the value taken. Note that there is a race condition whereby
   another process can put something in the 'MVar' after the take
   happens but before the put does.
 -}
 swapMVar :: NFData a => MVar a -> a -> IO a
-swapMVar mvar new = mask $ \_ -> do
+swapMVar !mvar new = seq (force new) $ mask $ \_ -> do
     old <- takeMVar mvar
-    putMVar mvar new
+    MV.putMVar mvar new
     return old
 
-{-|
-  'withMVar' is a safe wrapper for operating on the contents of an
-  'MVar'.  This operation is exception-safe: it will replace the
-  original contents of the 'MVar' if an exception is raised (see
-  "Control.Exception").
--}
-{-# INLINE withMVar #-}
--- inlining has been reported to have dramatic effects; see
--- http://www.haskell.org//pipermail/haskell/2006-May/017907.html
-withMVar :: NFData a => MVar a -> (a -> IO b) -> IO b
-withMVar m io = mask $ \unmask -> do
-    a <- takeMVar m
-    b <- Exception.catch (unmask (io a))
-            (\(e :: SomeException) -> do putMVar m a; throw e)
-    putMVar m a
-    return b
-
 {-|
   A safe wrapper for modifying the contents of an 'MVar'.  Like 'withMVar', 
   'modifyMVar' will replace the original contents of the 'MVar' if an
   exception is raised during the operation.
 -}
 {-# INLINE modifyMVar_ #-}
 modifyMVar_ :: NFData a => MVar a -> (a -> IO a) -> IO ()
-modifyMVar_ m io = mask $ \unmask -> do
-    a  <- takeMVar m
-    a' <- Exception.catch (unmask (io a))
-            (\(e :: SomeException) -> do putMVar m a; throw e)
-    putMVar m a'
+modifyMVar_ m io = MV.modifyMVar_ m $ io >=> \p -> force p `seq` return p
```

Comment:
Bang on m?
```suggestion
modifyMVar_ !m io = MV.modifyMVar_ m $ io >=> \p -> force p `seq` return p
```

Reply:
No need. `MV.modifyMVar_` does that for us.